main/src/library/Iterator.flix

```diff
@@ -14,179 +14,158 @@
  * limitations under the License.
  */
 
-pub enum Iterator[a: Type, r: Region] {
-    case Iterator(Iterator.Done[r], Iterator.Next[a, r])
+pub enum Iterator[a: Type, ef: Bool, r: Region] {
```

Comment:
I think this is a beautiful way to view iterators. There is the memory effects used by the iterator itself (r) and the user-effects (ef) collected by the user.

Reply:
Yes - effectful Iterators seem to form a (graded) monad - the set of effects possibly expanding as operations are chained together with `flatMap`.

I've changed the constructors `empty`, `singleton`, `range` and `repeat` to duplicate the region as the effect rather than them having a polymorphic effect through casting. I'm not 100% sure this is the right thing to do but it does make them simpler at the moment.